### PR TITLE
DataFrame: expose an object array as a data frame

### PR DESCRIPTION
**Pull request by ryantxu**
When working with data in the real world, it is often a list of objects.  We have a few methods to build a DataFrame from an object... but none feel great.  

Super naive implementations work great:
```
const frame = new MutableDataFrame({});
for(const row of data) {
   frame.add(row);
}
```

but when you have to manipulate the values it gets pretty weird:
```
const frame = new MutableDataFrame({});
for(const row of data) {
  const time = dateTime(row.time, 'YYYYMMDD').valueOf(); 
  frame.add({...row, time});
}
...
later set the `type` for the time field to time
```

This adds a function that lets you create a DataFrame directly from an Array -- interestingly it keeps the original data untouched and just looks thins up by index and property name.  It also has a conversion function so you can easily modify fields as needed:

```
const frame = new ArrayDataFrame( data );
frame.setFieldType( 'time', FieldType.time, (v) => dateTime(row.time, 'YYYYMMDD').valueOf() );
```

### CODE REVIEW

File: packages/grafana-data/src/vector/FunctionalVector.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  }
+
+  // Implement "iterable protocol"
+  [Symbol.iterator]() {
+    return this.iterator();
+  }
+
+  forEach(iterator: (row: T) => void) {
+    for (let i = 0; i < this.length; i++) {
```
Comment (scottlepp) [inline packages/grafana-data/src/vector/FunctionalVector.ts:22]: should be able to just iterate here:  for(const item of this)

File: packages/grafana-data/src/vector/FunctionalVector.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  // Implement "iterator protocol"
+  *iterator() {
+    for (let i = 0; i < this.length; i++) {
+      yield this.get(i);
+    }
+  }
+
+  // Implement "iterable protocol"
+  [Symbol.iterator]() {
+    return this.iterator();
```
Comment (scottlepp) [inline packages/grafana-data/src/vector/FunctionalVector.ts:18]: return vectorator(this)?  then remove the generator here.  seems like that should work

File: packages/grafana-data/src/vector/FunctionalVector.ts (hunk 1 above)
Comment (scottlepp) [inline packages/grafana-data/src/vector/FunctionalVector.ts:22]: return vectorator(this).forEach...   add forEach to vectorator?

File: packages/grafana-data/src/vector/FunctionalVector.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  forEach(iterator: (row: T) => void) {
+    for (let i = 0; i < this.length; i++) {
+      iterator(this.get(i));
+    }
+  }
+
+  map<V>(iterator: (item: T, index: number) => V) {
+    const acc: V[] = [];
```
Comment (scottlepp) [inline packages/grafana-data/src/vector/FunctionalVector.ts:28]: return vectorator(this).map...  change the vectorator map function to support index.

File: packages/grafana-data/src/vector/FunctionalVector.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    for (let i = 0; i < this.length; i++) {
+      acc.push(iterator(this.get(i), i));
+    }
+    return acc;
+  }
+
```
Comment (scottlepp) [inline packages/grafana-data/src/vector/FunctionalVector.ts:34]: filter would be nice.  i think you can just delegate to the vectorator

**Review by scottlepp — CHANGES_REQUESTED** [review]
Nice!  It seems the FunctionalVector and vectorator are a bit redundant.  Could you just delegate to the vectorator for the iterator and functional methods?  Ideally vecorator map would support index anyway the same way FunctionalVector does.

File: packages/grafana-data/src/vector/FunctionalVector.ts (hunk 2 above)
Comment (ryantxu) [reply packages/grafana-data/src/vector/FunctionalVector.ts:18]: :( i does not... all the other functins worked fine, but this one complains about type mismatch

File: packages/grafana-data/src/vector/FunctionalVector.ts (hunk 1 above)
Comment (ryantxu) [reply packages/grafana-data/src/vector/FunctionalVector.ts]: you got to let me keep for loops somewhere!

**Review by scottlepp — APPROVED** [review]

**Pull request merged**